Feature/entry sections

Mwexim opened the pull request:
Entry sections are sections that require a specific set of entries to act as some sort of configuration.

Skript's command syntax is the perfect example.
```skript
command /test:
    usage: &cUsage: /test
    permission: plugin.test
    permission message: &cYou lack the permission to do that.
    trigger:
        # Do stuff
```
We can immediately see the key features of the new `SectionConfiguration`.
- Optional and required 'option' entries that store a string value (like the permission or usage).
- Sections that store an entire `CodeSection` inside (like the trigger entry).

The `SectionConfiguration` class essentially makes the creation of these configuration sections easy, handling all possibilities, error messages and more.

This pull request is not completed yet as I still feel like some refinements should be made.

Mwexim committed:
Merge pull request #1 from SkriptLang/master

merge: merge of all commits until 2/1/2021

Commit:
✨ CategorySection

Commit:
✨ SectionConfiguration

WeeskyBDW reviewed (dismissed):
Look good to me

ItsTheSky commented:
OwO
Thank you a lot for adding this that fast!
It seems to work as same as for the olds one, so it will be easier to use.
However, could you maybe provide a way to directly parse input using the section ? Lemme explain:

Imagine we have this code:
```applescript
define website:
    port: 5030
    trigger:
        # </>
```

Instead of getting the `port` value directly, and then parsing it from our side, why not add a `Function` in the `addEntry()` method, for example:

```java
new SectionConfiguration()
        .addOption("port", false, input -> {
            // input is a String
            /*
             We could return an optional here:
             - If the optional is empty, then an error occurred and SectionConfiguration will return false
             - Else, it will attribute the parsed optional to this entry key and will return it when using getOption()
            */
            try {
                return Optional.of(Integer.parseInt(input));
            } catch (Exception ex) {
                logger.error("Invalid integer input: " + input);
                return Optional.empty();
            }
        });
```

Olyno commented:
```java
.addOption("port", false, input -> {
    return Integer.parseInt(input);
});
```
This should be better. Optionals are important but only skript-parser should deal with it mainly. When getting options, we simply return a ``Optional`` of the option's value.

Commit:
✨ EntryLoader (to customize the parsing process)

Mwexim commented:
While the latest commit does not add the proposed feature directly, it makes it possible to do so. I will push the last commit soon, which will add the proposed feature and make it even easier to handle.

Furthermore, I need to add Javadoc, a better explanation of what I'm doing and overall code improvements. The code is still somewhat all over the place but I'm way happier with what I created now than the old system. It's way more advanced than Skript's currently and allows developers to easily extend on the current system, which already provides many tools.

I'm currently not really asking for a review, I will do that when I'm done with the last changes.

Commit:
✨ LiteralLoader (to add literal values)

Commit:
📝 EntryLoader (Javadoc)

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
Merge branch 'master' of https://github.com/SkriptLang/skript-parser

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
Merge remote-tracking branch 'origin/master'

WeeskyBDW commented:
@Mwexim any news about that PR ?

Mwexim commented:
> @Mwexim any news about that PR ?

This is almost ready to be merged honestly. Just a quick sweep/refactoring is necessary to make it more user-friendly, as well as some Javadocs explaining how everything works.

TheLimeGlass commented on src/main/java/io/github/syst3ms/skriptparser/lang/entries/SectionConfiguration.java:
```diff
@@ -0,0 +1,125 @@
+		entries.add(new OptionLoader(false, key, false));
+		return this;
+	}
+
```
Optionals will be used frequently, so I think it'll be best to have a method for adding optionals.
```suggestion

    /**
     * Adds an optional entry to this {@link SectionConfiguration}
     * 
     * @param key The key of the entry.
     * @return This {@link SectionConfiguration} with the newly added entry for building.
     */
    public SectionConfiguration addOptional(String key) {
        entries.add(new OptionLoader(false, key, true));
        return this;
    }


    /**
     * Adds an optional list entry to this {@link SectionConfiguration}
     * 
     * @param key The key of the entry.
     * @return This {@link SectionConfiguration} with the newly added entry for building.
     */
    public SectionConfiguration addOptionalList(String key) {
        entries.add(new OptionLoader(true, key, true));
        return this;
    }

```

Mwexim replied:
I think I'm going to refactor the 'optional' option resulting in easier registration, because the current system is indeed a bit cumbersome.

TheLimeGlass commented on src/main/java/io/github/syst3ms/skriptparser/lang/entries/SectionConfiguration.java:
```diff
@@ -0,0 +1,125 @@
+import java.util.List;
+import java.util.Map;
+
+public class SectionConfiguration {
+	@Nullable
```
Systems used to do this a lot
```suggestion
public class SectionConfiguration {

	@Nullable
```

TheLimeGlass commented:
I think returning an Optional for getValue could be nice, as the entries may not be present, rather than null.

Also a cool feature would be allowing a pattern to be placed in the add entry methods, to allow for a pattern to be used to parse the key. like `(executor|command sender)` to allow for either `executor:` or `command sender:`

TheLimeGlass commented on src/main/java/io/github/syst3ms/skriptparser/lang/entries/SectionConfiguration.java:
```diff
@@ -0,0 +1,125 @@
+	}
+
+	@SuppressWarnings("unchecked")
+	public <T> T getValue(String key, Class<T> cls) {
```
When I had 
```
key: true
```
and used
```java
configuration.getValue("key", Boolean.class);
```
The returned value was always a String, so this would cause a class cast exception.
```
java.lang.ClassCastException: class java.lang.String cannot be cast to class java.lang.Boolean (java.lang.String and java.lang.Boolean are in module java.base of loader 'bootstrap')
```
This is when using a non literal. That should error or be caught. When using a literal entry it's fine.

This also happens when a type is not found. So if `configuration.getValue("key", AssuemdClass.class);` and AssumedClass is assumed to be loaded, but it isn't. It'll always return a String, where I think it should error stating that the type does not exist.

Mwexim replied:
This is indeed confusing.

Mwexim replied:
Added a descriptive error message containing information about this special use-case, but honestly, I'm not sure if I will keep adding these 'edge-case error messages'.

Mwexim committed:
Merge remote-tracking branch 'origin/master'

Mwexim committed:
Merge branch 'master' into feature/entry-sections

Mwexim committed:
Merge remote-tracking branch 'origin/master'

Mwexim committed:
Merge branch 'master' into feature/entry-sections

Mwexim committed:
SectionConfiguration: refactored a lot of code, fixed issues, improved error messages

Mwexim committed:
SectionLoader: small change

Mwexim commented:
So I refactored a lot of code, resulting in breaking changes. I'm sorry to do this abruptly but I just couldn't stand the old implementation. If you are just implementing this, the changes will be minimal as only some methods have changed. I also added a lot of Javadoc and more descriptive error messages to further help future developers along.
